File: lua/actor/print_server.lua
--[[
  print_server.lua
  Actor-model print server using coroutines + mailbox.

  Each "actor" is a coroutine with a message queue (table).
  Supports tell (fire-and-forget) and ask (request-response).
  Compare with: Python actor_print_server.py, Erlang gen_server
]]

local PrintServer = {}
PrintServer.__index = PrintServer

function PrintServer.new()
  local self = setmetatable({
    mailbox = {},
    counter = 0,
    running = true
  }, PrintServer)

  -- The actor runs as a coroutine processing messages
  self.actor = coroutine.create(function()
    self:run()
  end)

  return self
end

function PrintServer:run()
  while self.running do
    if #self.mailbox > 0 then
      local msg = table.remove(self.mailbox, 1)
      self.counter = self.counter + 1

      -- Simulate processing
      local output = string.format("[PrintServer] #%d printed: %s", self.counter, msg.text)
      print("  " .. output)

      if msg.reply_to then
        msg.reply_to(output)
      end
    end
    coroutine.yield()
  end
  print("  [PrintServer] shutting down")
end

function PrintServer:tell(text)
  table.insert(self.mailbox, { text = text, reply_to = nil })
end

function PrintServer:ask(text)
  local result
  local done = false
  table.insert(self.mailbox, {
    text = text,
    reply_to = function(reply)
      result = reply
      done = true
    end
  })
  -- Process until we get the reply
  while not done do
    coroutine.resume(self.actor)
  end
  return result
end

function PrintServer:step()
  coroutine.resume(self.actor)
end

function PrintServer:stop()
  self.running = false
  coroutine.resume(self.actor)
end

-- Demo
print("=== Actor: PrintServer Demo (Lua) ===\n")

local server = PrintServer.new()
print("  [PrintServer] started")

-- Fire-and-forget
server:tell("Hello, Actor World!")
server:tell("Printing document #1")
server:tell("Printing document #2")
server:step() -- process one batch

-- Request-response
local reply = server:ask("Urgent report")
print("  [Client] got reply: " .. reply)

server:step()

server:tell("Final document")
server:step()
server:stop()

print("\nPrintServer demo passed.")

File: kotlin/actor/PrintServer.kt
/**
 * PrintServer.kt
 *
 * Actor model using kotlinx.coroutines (send/receive on Channel).
 * Demonstrates tell (fire-and-forget) and ask (request-response).
 *
 * Compare with: Python actor_print_server.py, Erlang gen_server
 *
 * Compile: kotlinc -cp /usr/share/java/kotlinx-coroutines-core-1.0.1.jar \
 *                  -d out actor/PrintServer.kt
 * Run:     java -cp /usr/share/java/kotlinx-coroutines-core-1.0.1.jar:out \
 *                  PrintServerKt
 */

import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.launch
import kotlinx.coroutines.delay

class PrintServer(private val mailbox: Channel<String> = Channel(Channel.UNLIMITED)) {
    private var counter = 0
    private var running = true

    /** Process messages — call from a coroutine. */
    suspend fun process() {
        println("  [PrintServer] started")
        while (running) {
            val result = mailbox.receiveCatching()
            if (result.isClosed) break
            val msg = result.getOrNull() ?: break
            counter++
            delay(20)
            val output = "[PrintServer] #$counter printed: $msg"
            println("  $output")
        }
        println("  [PrintServer] shutting down")
    }

    /** Fire-and-forget: send a message, no reply expected. */
    fun tell(text: String) {
        mailbox.trySend(text)
    }

    fun stop() {
        running = false
        mailbox.close()
    }
}

fun main(args: Array<String>) = runBlocking {
    println("=== Actor: PrintServer Demo ===\n")

    val mailbox = Channel<String>(Channel.UNLIMITED)
    val server = PrintServer(mailbox)

    // Launch the actor processor in a background coroutine
    val job = launch {
        server.process()
    }

    // Fire-and-forget
    server.tell("Hello, Actor World!")
    server.tell("Printing document #1")
    server.tell("Printing document #2")

    delay(100)
    server.tell("Final document")
    delay(50)
    server.stop()

    job.join()
    println("\nPrintServer demo passed.")
}
